### PR TITLE
Add approve and decline api endpoints

### PR DESCRIPTION
**Pull request by christopappas**
- Adds approve and deny endpoints for licenses and coupons
- Hitting approval endpoint will hit license-manager/ecommerce to verify there are available licenses/coupons available

### CODE REVIEW

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:85]: Should these be HTTP_400_BAD_REQUESTs? Alternatively if you wanted to you could create a serializer and put the validation logic in validate(), which should automatically raise 400s.

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:164]: subsidy_request_uuid  -> subscription_plan_uuid in comments here

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:185]: nit: subscriptions remaining -> licenses remaining

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:193]: you could also use `next(status_count['count'] for status_count in status_counts if status_count['status'] == 'unassigned')`.

nit: should we move 'unassigned' to a variable in constants?

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:205]: I think this should be REQUESTS_ADMIN_ACCESS_PERMISSION since we only want to allow admins to approve.

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:228]: [clarification]: if the subsidy request has already been approved, will we just ignore them in the task?

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:246]: same permission as above

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:265]: we can't deny approved requests as well I believe so we need to handle that.

File: enterprise_access/apps/api/v1/views.py
Comment (long74100) [inline enterprise_access/apps/api/v1/views.py:335]: nit: could probably pull this part out into a helper?

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:164]: ty

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:193]: personally think is non-fancy way it's written now is more readable 🤷 

and sure, can move that to a constant

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:85]: yeah i can change to HTTP_400_BAD_REQUESTs

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:205]: gooood call

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:228]: we can, though it's probably better if we don't spin off a task at all. i can update this to keep track of which approval requests have already been approved so that we have the right set of things to run celery tasks for

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:335]: tis 1 if-statement that returns the response to the caller, if there were more bits i would consider this, but seems  too small imo

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:265]: i guess i figured once something is approved, the UI wouldn't give the option to deny it (and even if someone hit the API directly, it wouldn't change the state of the license, but this is probably the right thing to do, so i'll add it

Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:275]: added logic to prevent approved and pending approved requests from being denied 👍

File: enterprise_access/apps/api/v1/views.py
Comment (macdiesel) [reply enterprise_access/apps/api/v1/views.py:252]: I'm not going to mark any more of these, but please do take a look at refactoring on the docs.

File: enterprise_access/apps/api_client/ecommerce_client.py
Comment (macdiesel) [inline enterprise_access/apps/api_client/ecommerce_client.py:9]: Same question for the `LicenseManagerAPIClient`: Can we set a client side timeout of these requests to ~45 seconds? This way this code base can catch the timeout and do something sane without instead of leaving the users browser spinning or returning a generic 500 error?

File: enterprise_access/apps/api_client/ecommerce_client.py
Comment (macdiesel) [inline enterprise_access/apps/api_client/ecommerce_client.py:18]: I assume this is copy pasta, that said, this is what we should shoot for docstring wise.

File: enterprise_access/apps/api_client/license_manager_client.py
Comment (macdiesel) [inline enterprise_access/apps/api_client/license_manager_client.py:14]: Should we set these as config options instead of inline?

File: enterprise_access/apps/api_client/tests/test_ecommerce_client.py
Comment (macdiesel) [inline enterprise_access/apps/api_client/tests/test_ecommerce_client.py:12]: Would suggest adding some error case tests.

File: enterprise_access/settings/devstack.py
Comment (macdiesel) [inline enterprise_access/settings/devstack.py:62]: Are there devstack URLs that would actually work we could put in here?

**Review by macdiesel — COMMENTED** [review]
Left a few thoughts and ideas. I did not review the tests for the views.

File: enterprise_access/settings/devstack.py
Comment (christopappas) [reply enterprise_access/settings/devstack.py:62]: actually kinda curious, if you run tests locally, does it say it's using devstack.py as the env? i found locally that test env was being ignored (and using devstack instead). so i put those there to get tests to work. but yeah, i can change these once i get the tests running against the right env

Comment (macdiesel) [reply enterprise_access/settings/devstack.py:62]: I did not try running, only read the code.

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:229]: different approach: on deny/approve endpoints, we filter the uuids handed to us, and only actually act on things that are appropriate to act on

so if you had 5 UUIDs for approval, but 1 is already approved, we just approve 4 and do not error. and same for denying, if you go to deny a thing that is already denied, we simply do nothing instead of error

thoughts?

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:165]: weird that pydocstyle didn't complain, i will (begrudgingly) change this

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: added in tries here and with the other client. please take a look at the language i use in the error message. happy to change it to whatever

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: updated

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: updated

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: updated

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: updated

File: enterprise_access/apps/api_client/license_manager_client.py
Comment (christopappas) [reply enterprise_access/apps/api_client/license_manager_client.py:14]: probably not. LICENSE_MANAGER_URL makes sense  to be in settings, but the expected paths from that service should always be the same. if we ever update LM or ecommerce to use api/v2, then we can make a change here then

Comment (christopappas) [reply enterprise_access/apps/api_client/license_manager_client.py:14]: (in particular cause these paths are used nowhere but here. and if it's any consolation, that's how it is in LM)

File: enterprise_access/settings/devstack.py
Comment (christopappas) [reply enterprise_access/settings/devstack.py:62]: alrighty, i'm inclined to file a follow up ticket/make a follow up PR to change this value in devstack.py. this PR has enough stuff in it ><'

agreed though that devstack.py and test.py should have different values

File: enterprise_access/apps/api_client/license_manager_client.py
Comment (macdiesel) [reply enterprise_access/apps/api_client/license_manager_client.py:14]: 👍

File: enterprise_access/settings/devstack.py
Comment (macdiesel) [reply enterprise_access/settings/devstack.py:62]: Follow up ticket on this is fine.

File: enterprise_access/apps/api/v1/views.py
Comment (macdiesel) [reply enterprise_access/apps/api/v1/views.py:251]: hrm.... I wonder what the industry at large thinks about stuff like this? IMO to be semantically correct it should either be all or nothing. That said I'm not sure that provides the best experience either. It does seem OK to let the API return a 200 in the case when you have n requests but 1+ of those are already in the state you are attempting to set. That seems reasonable. 

What about partial applies? IF only one state is able to be changed, but the others not, what should we do there? Return a partial result? A partial error? for simplicity up front it would likely be better for us to make the failure case all or nothing. But I think we are already set for this as an exception raised would cause the atomic block to rollback.

I think that's my long winded way of saying I think this is ok....

File: enterprise_access/apps/api/v1/views.py
Comment (macdiesel) [inline enterprise_access/apps/api/v1/views.py:201]: Maybe something more specific to leave ourselves a better breadcrumb when/if this happens?  Perhaps unable to assign (code/licenses) at this time. Please try again later or reach out ..... blah blah blah? Thoughts?

File: enterprise_access/settings/devstack.py
Comment (christopappas) [reply enterprise_access/settings/devstack.py:62]: ticket created

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:201]: yeah that sounds reasonable

Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]:
```                
'We were unable to approve/deny the selected coupon requests at this time. '
'Please try again, or reach out to customer support if you '
'continue to experience issues. '
```

and then replace coupon with license for the other 👍

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py]: updated

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:194]: updated

File: enterprise_access/apps/api_client/tests/test_ecommerce_client.py
Comment (christopappas) [reply enterprise_access/apps/api_client/tests/test_ecommerce_client.py:12]: so this (and the license manager) client is importing from `BaseOAuthClient` which, imports from: `from edx_rest_api_client.client import OAuthAPIClient`, because the custom logic we add is really just making the URL and then doing a GET, I personally don't think tests here are necessary (I'm making the assumption that the package we're importing from does testing of its own)

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:255]: cool. i did partially mistype:

an admin tries to deny 4 requests, 1 of which has already been approved. error? or simply ignore the 1 approved and  deny the 3-not-yet-denied?

File: enterprise_access/apps/api_client/ecommerce_client.py
Comment (christopappas) [reply enterprise_access/apps/api_client/ecommerce_client.py:9]: Added in 45 second timeouts for both clients

File: enterprise_access/apps/api/v1/views.py
Comment (macdiesel) [reply enterprise_access/apps/api/v1/views.py:256]: > an admin tries to deny 4 requests, 1 of which has already been approved. error?

Yes I do think this is an error case. That said I don't believe/am hopeful the UI wouldn't allow this state to be submitted.

File: enterprise_access/apps/api_client/tests/test_ecommerce_client.py
Comment (macdiesel) [reply enterprise_access/apps/api_client/tests/test_ecommerce_client.py:13]: I was thinking about the default error cases: timeout, host not found, stuff like that. If this stuff is already covered please proceed as completed.

File: enterprise_access/apps/api/v1/views.py
Comment (macdiesel) [reply enterprise_access/apps/api/v1/views.py]: Perhaps "Please try again in a little while, or ...

Alena, ideas on language for this error?

File: enterprise_access/apps/api/v1/views.py
Comment (christopappas) [reply enterprise_access/apps/api/v1/views.py:260]: good question. i'm actually going to refactor this a bit to make it a little more straight forward for hooking up celery task stuff

File: enterprise_access/apps/api/constants.py
Comment (christopappas) [reply enterprise_access/apps/api/constants.py]: updated

File: enterprise_access/settings/test.py
Comment (long74100) [inline enterprise_access/settings/test.py:26]: nit: new line

**Review by long74100 — APPROVED** [review]
🐳

**Review by macdiesel — APPROVED** [review]

**Pull request merged**
